CASSANDRA-17639 May 2022 blog "The Path to Green CI" (#130)
apache/cassandra-website: site-content/source/modules/ROOT/pages/blog/The-Path-to-Green-CI.adoc

= The Path to Green CI
:page-layout: single-post
:page-role: blog-post
:page-post-date: May, 19 2022
:page-post-author: Josh McKenzie
:description: Testing Apache Cassandra
:keywords:

:!figure-caption:

.Image credit: https://unsplash.com/@hasanalmasi[Hasan Almasi on Unsplash^]
image::blog/the-path-to-green-ci-unsplash-hasan-almasi.jpg[the path to Green CI]

As we approach the Cassandra 4.1 GA release, it’s a great time to stop and reflect on some of our development history, the past year since we released 4.0, and where we’re headed in the future. In this blog post, we’re going to limit our focus to testing Cassandra as that’s been a huge focus in the 4.0+ time frame.

=== The Numbers

> But don’t just take my word for it - let’s start with some numbers! We’re going to use “lines of code” as a loose proxy for “where we’re spending our time”. This is definitely a fraught metric, but as _one_ way of viewing things it paints a pretty interesting picture, consistent with many of our intuitions.

Reviewer: The acronym `LoC` gets used further down so I'm defining it here as `lines of code`:
```suggestion
But don’t just take my word for it - let’s start with some numbers! We’re going to use “lines of code” (LoC) as a loose proxy for “where we’re spending our time”. This is definitely a fraught metric, but as _one_ way of viewing things it paints a pretty interesting picture, consistent with many of our intuitions.
```

Below is a list for our past three major releases and how much raw code exists in the following:

> 1. src/java: raw database code

Reviewer: ```suggestion
1. `src/java`: raw database code
```

> 2. test/unit: unit testing code

Reviewer: ```suggestion
2. `test/unit`: unit testing code
```

> 3. test/distributed: distributed tests in java code

Reviewer: ```suggestion
3. `test/distributed`: distributed tests in java code
```

> 4. The cassandra-dtest repo .py test files: python distributed tests using https://github.com/riptano/ccm[ccm^]

Reviewer: ```suggestion
4. The `cassandra-dtest` repo .py test files: python distributed tests using https://github.com/riptano/ccm[ccm^]
```

> [cols=3*]

Reviewer: Mark the first row as the table header:
```suggestion
[%header,cols=3*]
```

|=======

> |||*% total*

Reviewer: ```suggestion
|&nbsp; |Lines |% total
```

> 

Reviewer: ```suggestion
```

> |*3.0 Total:*|*316,262*|100.00%

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

|src|181,871|57.51%

> 

Reviewer: ```suggestion
```

|junit|100,284|31.71%

> 

Reviewer: ```suggestion
```

|jdtest|9,225|2.92%

> 

Reviewer: ```suggestion
```

|pdtest|24,882|7.87%

> 

Reviewer: ```suggestion
|*3.0 Total*|*316,262*|100.00%
```

|All tests|134,391|42.49%
|=======

> 

Reviewer: Add a blank line:
```suggestion

{sp} +

```

> [cols=4*]

Reviewer: Mark the first row as the table header:
```suggestion
[%header,cols=4*]
```

|=====

> |||*% total*|*% change since 3.0*

Reviewer: ```suggestion
|&nbsp; |Lines |% total |% change since 3.0
```

> 

Reviewer: Delete blank line:
```suggestion
```

> |*4.0 Total:*|495,566|100.00%|156.69%

Reviewer: ```suggestion
```

> 

Reviewer: Delete blank line:
```suggestion
```

|src|261,825|52.83%|143.96%

> 

Reviewer: Delete blank line:
```suggestion
```

|junit|172,672|34.84%|172.18%

> 

Reviewer: Delete blank line:
```suggestion
```

|jdtest|21,769|4.39%|235.98%

> 

Reviewer: Delete blank line:
```suggestion
```

|pdtest|39,300|7.93%|157.95%

> 

Reviewer: ```suggestion
|*4.0 Total*|*495,566*|100.00%|156.69%
```

|All tests|233,741|47.17%|
|=======

> |=====

Reviewer: Delete unnecessary table markdown:
```suggestion
```

> 

Reviewer: Add a blank line:
```suggestion

{sp} +

```

> [cols=4*]

Reviewer: Mark the first row as the table header:

```suggestion
[%header,cols=4*]
```

|=======
|||*% total*|*% change since 4.0*

> 

Reviewer: Delete blank line:
```suggestion
```

> |*4.1 Total:*|566,127|100.00%|114.24%

Reviewer: ```suggestion
```

> 

Reviewer: Delete blank line:
```suggestion
```

|src|297,685|52.58%|113.70%

> 

Reviewer: Delete blank line:
```suggestion
```

|junit|197,231|34.84%|114.22%

> 

Reviewer: Delete blank line:
```suggestion
```

|jdtest|31,306|5.53%|143.81%

> 

Reviewer: Delete blank line:
```suggestion
```

|pdtest|39,905|7.05%|101.54%

> 

Reviewer: ```suggestion
|*4.1 Total*|*566,127*|100.00%|114.24%
```

We can also infer that our new code addition to the database has _accelerated_ in the past year, as the delta from 4.0-4.1 represents a time frame of one calendar year and roughly 38k LoC net add vs. the 5.5 year gap between 3.0 and 4.0 with a net add of ~80k LoC. While the database also had a release line of 3.1-3.11 introducing new features and tests during this time window; for the sake of this analysis we’re only considering major traditional point releases.

So what does all this mean for us working on and depending on the project? It means that with the release of 4.1, _we have 10% more code just *unit* testing the database than we had in the entire database in 3.0_. It means that new development on Cassandra is accelerating. It also means we’re constantly moving the goalposts on what’s required to keep https://ci-cassandra.apache.org/[Green (passing) CI (continuous integration)^].

=== Keeping it Green

> We all know Cassandra is an incredibly complex piece of software. The power to scale up linearly to petabytes of data on hundreds of machines, with zero downtime, in a masterless single logical cluster, where machines can drop out and in, https://cassandra.apache.org/doc/latest/cassandra/operating/hints.html[hint], https://cassandra.apache.org/doc/latest/cassandra/operating/read_repair.html[heal], and https://cassandra.apache.org/doc/latest/cassandra/operating/repair.html[repair] simply cannot be implemented without a significant amount of code, infrastructure, and testing to ensure it works as expected. Further, as we support hot upgrades with zero downtime between versions with mixed version clusters running, we have a strong commitment to backwards compatibility in the mix.

Reviewer: Replace all hardcoded `/doc` links:

```suggestion
We all know Cassandra is an incredibly complex piece of software. The power to scale up linearly to petabytes of data on hundreds of machines, with zero downtime, in a masterless single logical cluster, where machines can drop out and in, link:/doc/latest/cassandra/operating/hints.adoc[hint], link:/doc/latest/cassandra/operating/read_repair.adoc[heal], and link:/doc/latest/cassandra/operating/repair.adoc[repair] simply cannot be implemented without a significant amount of code, infrastructure, and testing to ensure it works as expected. Further, as we support hot upgrades with zero downtime between versions with mixed version clusters running, we have a strong commitment to backwards compatibility in the mix.
```
/cc @dtopdontstop , @cthornett

We always drive down to stable 0 test failures at a GA release, however what we refer to as “flaky tests” sneak back into our suite over time. Let’s take a recent example, https://nightlies.apache.org/cassandra/ci-cassandra.apache.org/job/Cassandra-trunk/1112/[build run 1112^] on trunk (effectively Cassandra 4.1 pre-alpha).

19 test failures! https://nightlies.apache.org/cassandra/ci-cassandra.apache.org/job/Cassandra-trunk/1112/testReport/[Out of an entire suite of 49,704 tests makes that a 99.96% pass rate^]. Nobody wants a database that works 99.96% of the time, however, and that’s assuming we have 100% test coverage of not just all our code but also all possible combinations of state, a problem so daunting some contributors are https://issues.apache.org/jira/browse/CASSANDRA-15348[pushing the bleeding edge of the state of the art of distributed database testing^].

Burning down less than 20 flaky tests to get our release out between freeze and our goal for release, an eight-week window, is quite doable, so why not just continue to float along with a low number of test failures? Well, it gets more complicated when we look at _where_ we run our tests.

=== Circle vs. Jenkins

> As https://cassandra.apache.org/_/development/testing.html[we outline in our contributor guide on testing], tests can both be run on https://ci-cassandra.apache.org/[Apache Jenkins infrastructure^] or on https://github.com/apache/cassandra/tree/cassandra-4.1/.circleci[CircleCI^]. The primary difference between these two systems are runtime, cost, and resources allocated to each individual test. While some contributors have access to paid CircleCI accounts that allow them to dedicate more resources to their test runs and shorten feedback loops, this is an open source volunteer project and our canonical CI is the Apache Jenkins infrastructure.

Reviewer: @dtopdontstop / @cthornett , this should have been caught at draft review stage:
```suggestion
As xref:/development/testing.adoc[we outline in our contributor guide on testing], tests can both be run on https://ci-cassandra.apache.org/[Apache Jenkins infrastructure^] or on https://github.com/apache/cassandra/tree/cassandra-4.1/.circleci[CircleCI^]. The primary difference between these two systems are runtime, cost, and resources allocated to each individual test. While some contributors have access to paid CircleCI accounts that allow them to dedicate more resources to their test runs and shorten feedback loops, this is an open source volunteer project and our canonical CI is the Apache Jenkins infrastructure.
```
